### ISSUE TITLE
SetFaceTemplate not working

### ISSUE THREAD

**Issue by luiscosta15** [MxLabs/Anviz#45]
Hello,

I'm trying to use the 'SetFaceTemplate' function but i'm always getting the following error: Payload too big

I'm always sending a byte[2056], which was returned when i called the function 'GetFaceTemplate' with a specific user.

Can you help me?

**Comment by JohnKiller** [comment]
Can you post a small reproducible example?

**Comment by luiscosta15** [comment]
var manager = new AnvizManager();
            using (var device = await manager.Connect(deviceHost))
            {
                device.DeviceError += (s, e) => throw e;

                //Get the face template of user number 3
                var faceBytes = await device.GetFaceTemplate(3);

                //Removes the user 3 from the device
                await device.DeleteEmployeesData(3);

                //Trying to associate the face template with user number 1
                //This will throw an exception with the message "Payload too big"
                await device.SetFaceTemplate(1, faceBytes);
            }

**Comment by luiscosta15** [comment]
maybe i should explain my problem in a little bit more detail. 

i need to install several devices, with the same group of users, without having to add all the users, one by one, and registers all of theirs faces, and repeat the same process for each device.

i trying to do the following steps:

1. send all of the users data to one device;
2. for each user, send the face template through web, or set the face template int the device
3. for other devices, get all of the users of the 'pivot' device, send them all to the other devices, and for each user get the face template of the pivot device and set it on other devices

**Comment by ivigo91** [comment]
Hello, 

I am using Facepass 7 device and I found the same problem (payload too big) but specifically in my case, using 'enrollfingertemplate'.

I am apply the February 25th sample, but without modifyng anything just the line in 'program.cs' where I call 'setfacetemplate' instead 'setfingertemplate'.

From what I've read, that version supports anviz facepass7 but i can't record a faceand save it.

Also,I've tried call 'getfacetemplate' of a save face from the device but where I try 'setfacetemplate', I can´t do it because I also find the same error that Luiscosta15.

I do the same thing with various EP300 devices and it works.
Can you help me?Thanks

**Comment by JohnKiller** [comment]
Hi,
The error is a sanity check located at this line:

https://github.com/MxLabs/Anviz/blob/18949d31e07a9bb37603df3861360e81cb79a58a/Anviz.SDK/Commands/Command.cs#L23-L27

This follows the documentation, in which is stated that the DATA field ranges from 0 to 400 bytes.

Can you try commenting out this line and see if it works? Maybe the payload can be made bigger.

**Comment by ivigo91** [comment]
Thanks for your answer.

I have commented the IF lines, but I have found 3 problems:

1- Error in program.cs:  device.DeviceError += (s, e) => throw e; (commented)
2-  Error in program.cs: Console.WriteLine($"Employee {dict[rec.UserCode]} at {rec.DateTime.ToLongDateString()} {rec.DateTime.ToLongTimeString()}");  (commented)
3- There is no error but neither the employee nor the face is saved in device.

**Comment by JohnKiller** [comment]
Unfortunately I don't have that device to test the code. I suspect that the wire protocol may be completely different.

Just to clarify - does the other "simpler" functions like setting and reading the device date and time work ok?

**Comment by ivigo91** [comment]
The other functions like get time or get ip for example, works.

I have tried both wifi and wire with the same result.

Anyway, I understand that the problem is with the face template in 'enrollfingertemplate' method, since it seems that it is not compatible with the setface method

**Comment by luiscosta15** [comment]
The only functions that does not seem to work are setfacetemplate and enrollfingertemplate. all the other functions that i have tested are working perfectly.

I've also tried to comment the line but i still got the same result

**Comment by Meswy** [comment]
Hi guys,
I have the same situation. I tried convert all code to vb  but the try catch give me the same error ( Payload too big ). Some update for the problem? thanks.

**Comment by JohnKiller** [comment]
As I said in my previous commend, I don't own that particular device.
If someone could provide a wireshark capture of the commands using another tool that works, I will implement it in this library

**Comment by Meswy** [comment]
I think the problem may be related to the type of connection and permissions. Let me explain, I could not connect Facepass 7 in client mode with Crosschex Standard, only in server mode. In server mode, if you register a face and load the user to Facepass7, it does not save it, it deletes all users, just like when you save a user's face with the setFaceTemplate method commenting on the "if" that contains the variable "Payload <400" . On the other hand, if you save the face and click on the "Set permissions" option on top of a user, it loads the user and his template correctly. I will continue investigating, if I have any news I will write it.

Has anyone been able to connect to the device in server mode with this library?

Thx.

**Comment by Meswy** [comment]
I got the Facepass7 to save the user and his facetemplate. I paste the example code .net:

                var fp = await device.EnrollFingerprint (5); // number of employee ID, Example 5
                var employee = new Anviz.SDK.Responses.UserInfo (5, "name of employee");
                await device.SetEmployeesData (employee);
                await device.SetFaceTemplate (employee.Id, fp);

Then i comment the if that contains payload<400 in metod "BuildPayload".

In summary you have to change the order, first the template is created with the user identifier and then you create the user and assign the template to it.

Can u upload the library?

Thx.

**Comment by JohnKiller** [comment]
Thank you very much @Meswy !
@luiscosta15 can you please test Meswy solution?

**Comment by Meswy** [comment]
Hi,

Is it possible to update the changes in the library?

1- comment the if that contains payload<400 in metod "BuildPayload"
2- comment in metod "EnrollFingerprintCommand", await DeviceStream.SendCommand(new EnrollFingerprintCommand(DeviceId, employeeID, false)); (does the enrollface procedure twice)

Thx

**Comment by JohnKiller** [comment]
@Meswy 

https://github.com/MxLabs/Anviz/blob/75355cd04719056b18301d67a9d8543a7f537126/Anviz.SDK/Commands/EnrollFingerprintCommand.cs#L26

You can now set `verifyCount` to 1. Expect a release in a couple of hours

**Comment by Meswy** [comment]
Thank you very much @JohnKiller
